e2e/test/iothub/service/DirectMethodE2ETests.cs

```diff
@@ -0,0 +1,127 @@
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using FluentAssertions.Specialized;
+using Microsoft.Azure.Devices.E2ETests.Helpers;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.Azure.Devices.Client;
+using System.Reflection;
+using Microsoft.Identity.Client;
+
+namespace Microsoft.Azure.Devices.E2ETests.iothub.service
+{
+    [TestClass]
+    [TestCategory("E2E")]
+    [TestCategory("IoTHub")]
+    public class DirectMethodE2ETests : E2EMsTestBase
+    {
+        private readonly string _devicePrefix = $"{nameof(DirectMethodE2ETests)}_";
+        private readonly string _modulePrefix = $"{nameof(DirectMethodE2ETests)}_Module_";
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_InvokeAsync_DeviceDoesNotExist()
+        {
+            // arrange
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod");
+
+            // act
+            Func<Task> act = async () => await serviceClient.DirectMethods.InvokeAsync("someNonexistentDevice", methodInvocation);
+
+            // assert
+
+            ExceptionAssertions<IotHubServiceException> errorContext = await act.Should().ThrowAsync<IotHubServiceException>();
+            errorContext.And.StatusCode.Should().Be(System.Net.HttpStatusCode.NotFound);
+            errorContext.And.IsTransient.Should().BeFalse();
+            errorContext.And.ErrorCode.Should().Be(IotHubServiceErrorCode.DeviceNotFound);
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_InvokeAsycn_ModuleDoesNotExist()
+        {
+            // arrange
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+
+            // ensure device exists but module does not
+            await using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(_devicePrefix);
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod");
+
+            // act
+            Func<Task> act = async () => await serviceClient.DirectMethods.InvokeAsync(testDevice.Id, "someNonexistentModule", methodInvocation);
+
+            // assert
+            ExceptionAssertions<IotHubServiceException> errorContext = await act.Should().ThrowAsync<IotHubServiceException>();
+            errorContext.And.StatusCode.Should().Be(System.Net.HttpStatusCode.NotFound);
+            errorContext.And.IsTransient.Should().BeFalse();
+            errorContext.And.ErrorCode.Should().Be(IotHubServiceErrorCode.ModuleNotFound);
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_DeviceOnline_DeviceNotSubscribedToDirectMethods()
+        {
+            // arrange
+            // device is online but device client is not explicitly opened
```

Comment:
Did you mean that the device exists but isn't explicitly opened = isn't online? A device being online would mean that it is open = connected to IoT Hub service.